From f1159a335016702799d40d72b38d8879d5da01f0 Mon Sep 17 00:00:00 2001
From: Martin Fick <mfick@codeaurora.org>
Date: Mon, 5 Dec 2011 16:27:27 -0700
Subject: [PATCH] Support [[user: <id>]] in comments

Add support for embedding user ids in change comments.
This is useful when Gerrit needs to comment about users
other than the user performing the action creating the
comment.  The user id is embedded with a wiki like
syntax:

  [[user: <id>]]

The comment rendering can now interpret this syntax and
will replace it with the current name of the user, or
their current email if that is not yet set.

The reason the user id is embedded is to allow Gerrit
to display the user using their current information, not
the information at the time the comment was made (that is
the case anywhere else in the UI where a user is
referenced).  Since a comment is free form text and could
conceivably have multiple users referenced this way in a
single comment, it is not really feasible to store the
user id via a database field.

This feature is not really meant for use by users, but no
measures are taken to prevent users from creating text in
this format which would render as a user.

Change-Id: I1e51e60431a69f702b75e89ac8e99b3b17fb3e8a
---
 .../gerrit/common/data/AccountInfoCache.java       |    5 +++
 .../google/gerrit/client/changes/ChangeScreen.java |    2 +-
 .../gerrit/client/ui/CommentLinkProcessor.java     |    1 +
 .../com/google/gerrit/client/ui/CommentPanel.java  |   29 ++++++++++++++-
 .../rpc/changedetail/ChangeDetailFactory.java      |   13 +++++++
 .../com/google/gerrit/reviewdb/ChangeMessage.java  |   38 +++++++++++++++++++-
 6 files changed, 85 insertions(+), 3 deletions(-)

diff --git a/gerrit-common/src/main/java/com/google/gerrit/common/data/AccountInfoCache.java b/gerrit-common/src/main/java/com/google/gerrit/common/data/AccountInfoCache.java
index cf5f20b..4d780e9 100644
--- a/gerrit-common/src/main/java/com/google/gerrit/common/data/AccountInfoCache.java
+++ b/gerrit-common/src/main/java/com/google/gerrit/common/data/AccountInfoCache.java
@@ -19,6 +19,7 @@ import com.google.gerrit.reviewdb.Account;
 import java.util.Collections;
 import java.util.HashMap;
 import java.util.Map;
+import java.util.Set;
 
 /** In-memory table of {@link AccountInfo}, indexed by {@link Account.Id}. */
 public class AccountInfoCache {
@@ -76,4 +77,8 @@ public class AccountInfoCache {
     assert this != EMPTY;
     accounts.putAll(other.accounts);
   }
+
+  public Set<Account.Id> getAccountIds() {
+    return accounts.keySet();
+  }
 }
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/ChangeScreen.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/ChangeScreen.java
index ed0f2f2..0088151 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/ChangeScreen.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/ChangeScreen.java
@@ -389,7 +389,7 @@ public class ChangeScreen extends Screen {
       }
 
       final CommentPanel cp =
-          new CommentPanel(author, msg.getWrittenOn(), msg.getMessage());
+          new CommentPanel(author, msg.getWrittenOn(), msg.getMessage(), accts);
       cp.setRecent(isRecent);
       cp.addStyleName(Gerrit.RESOURCES.css().commentPanelBorder());
       if (i == msgList.size() - 1) {
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/ui/CommentLinkProcessor.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/ui/CommentLinkProcessor.java
index f5ef2f4..be3c800 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/ui/CommentLinkProcessor.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/ui/CommentLinkProcessor.java
@@ -27,6 +27,7 @@ import java.util.List;
 public class CommentLinkProcessor {
   public static SafeHtml apply(SafeHtml buf) {
     try {
+
       return buf.replaceAll(Gerrit.getConfig().getCommentLinks());
 
     } catch (RuntimeException err) {
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/ui/CommentPanel.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/ui/CommentPanel.java
index 249c70c..fedce14 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/ui/CommentPanel.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/ui/CommentPanel.java
@@ -17,6 +17,9 @@ package com.google.gerrit.client.ui;
 import com.google.gerrit.client.FormatUtil;
 import com.google.gerrit.client.Gerrit;
 import com.google.gerrit.common.data.AccountInfo;
+import com.google.gerrit.common.data.AccountInfoCache;
+import com.google.gerrit.reviewdb.Account;
+import com.google.gerrit.reviewdb.ChangeMessage;
 import com.google.gwt.event.dom.client.BlurEvent;
 import com.google.gwt.event.dom.client.BlurHandler;
 import com.google.gwt.event.dom.client.ClickEvent;
@@ -44,6 +47,7 @@ import com.google.gwtexpui.safehtml.client.SafeHtml;
 import com.google.gwtexpui.safehtml.client.SafeHtmlBuilder;
 
 import java.util.Date;
+import java.util.Map;
 
 public class CommentPanel extends Composite implements HasDoubleClickHandlers,
     HasFocusHandlers, FocusHandler, HasBlurHandlers, BlurHandler {
@@ -57,9 +61,14 @@ public class CommentPanel extends Composite implements HasDoubleClickHandlers,
   private boolean recent;
 
   public CommentPanel(final AccountInfo author, final Date when, String message) {
+    this(author, when, message, null);
+  }
+
+  public CommentPanel(final AccountInfo author, final Date when, String message,
+      AccountInfoCache accountInfoCache) {
     this();
 
-    setMessageText(message);
+    setMessageText(message, accountInfoCache);
     setAuthorNameText(FormatUtil.name(author));
     setDateText(FormatUtil.shortFormat(when));
 
@@ -110,10 +119,18 @@ public class CommentPanel extends Composite implements HasDoubleClickHandlers,
   }
 
   protected void setMessageText(String message) {
+    setMessageText(message, null);
+  }
+
+  protected void setMessageText(String message, AccountInfoCache accountInfoCache) {
     if (message == null) {
       message = "";
     } else {
       message = message.trim();
+      if (accountInfoCache != null) {
+        message = formatUsers(message, ChangeMessage.formatUsersRE(
+            accountInfoCache.getAccountIds()), accountInfoCache);
+      }
     }
 
     messageSummary.setText(summarize(message));
@@ -122,6 +139,16 @@ public class CommentPanel extends Composite implements HasDoubleClickHandlers,
     SafeHtml.set(messageText, buf);
   }
 
+  private static String formatUsers(String message,
+      Map<Account.Id,String> formatsById, AccountInfoCache accountInfoCache) {
+
+    for (Account.Id id : formatsById.keySet()) {
+      message = message.replaceAll(formatsById.get(id),
+          FormatUtil.nameEmail(accountInfoCache.get(id)));
+    }
+    return message;
+  }
+
   public void setAuthorNameText(final String nameText) {
     header.setText(0, 0, nameText);
   }
diff --git a/gerrit-httpd/src/main/java/com/google/gerrit/httpd/rpc/changedetail/ChangeDetailFactory.java b/gerrit-httpd/src/main/java/com/google/gerrit/httpd/rpc/changedetail/ChangeDetailFactory.java
index 6529f0f..e04b868 100644
--- a/gerrit-httpd/src/main/java/com/google/gerrit/httpd/rpc/changedetail/ChangeDetailFactory.java
+++ b/gerrit-httpd/src/main/java/com/google/gerrit/httpd/rpc/changedetail/ChangeDetailFactory.java
@@ -57,6 +57,8 @@ import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
 
 /** Creates a {@link ChangeDetail} from a {@link Change}. */
 public class ChangeDetailFactory extends Handler<ChangeDetail> {
@@ -188,9 +190,20 @@ public class ChangeDetailFactory extends Handler<ChangeDetail> {
     detail.setMessages(msgList);
     for (final ChangeMessage m : detail.getMessages()) {
       aic.want(m.getAuthor());
+      aic.want(getChangeMessageAccountIds(m));
     }
   }
 
+  private static Set<Account.Id> getChangeMessageAccountIds(ChangeMessage msg) {
+    Set<Account.Id> ids = new HashSet();
+    Pattern p = Pattern.compile(ChangeMessage.formatUserRE(null));
+    Matcher m = p.matcher(msg.getMessage());
+    while (m.find()) {
+      ids.add(new Account.Id(Integer.valueOf(m.group(1))));
+    }
+    return ids;
+  }
+
   private void load() throws OrmException, NoSuchChangeException {
     if (detail.getChange().getStatus().equals(Change.Status.NEW) && testMerge) {
       ChangeUtil.testMerge(opFactory, detail.getChange());
diff --git a/gerrit-reviewdb/src/main/java/com/google/gerrit/reviewdb/ChangeMessage.java b/gerrit-reviewdb/src/main/java/com/google/gerrit/reviewdb/ChangeMessage.java
index d8f1dd4..c9cee02 100644
--- a/gerrit-reviewdb/src/main/java/com/google/gerrit/reviewdb/ChangeMessage.java
+++ b/gerrit-reviewdb/src/main/java/com/google/gerrit/reviewdb/ChangeMessage.java
@@ -19,6 +19,10 @@ import com.google.gwtorm.client.StringKey;
 
 import java.sql.Timestamp;
 
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
 /** A message attached to a {@link Change}. */
 public final class ChangeMessage {
   public static class Key extends StringKey<Change.Id> {
@@ -125,4 +129,36 @@ public final class ChangeMessage {
   public void setPatchSetId(PatchSet.Id id) {
     patchset = id;
   }
-}
\ No newline at end of file
+
+  public static Map<Account.Id,String> formatUsersRE(Set<Account.Id> ids) {
+    Map<Account.Id,String> formats =
+        new HashMap<Account.Id,String>(ids.size());
+    for (Account.Id id : ids) {
+      formats.put(id, formatUserRE(id));
+    }
+    return formats;
+  }
+
+  /** Get the RexExp to find a user entry in a ChangeMessage
+    * created by formatUser(). Supplying a null Account.Id
+    * will get the RexExp needed to find any embedded user
+    * entry. Group 1 of the RegExp match will represent the
+    * matched Account.Id.
+    */
+  public static String formatUserRE(Account.Id aid) {
+    String id = aid != null ? aid.toString() : "[0-9]+";
+    return "\\[\\[user: *(" + id + ")\\]\\]";
+  }
+
+  /** Format a user as pseudo wiki link, based on their ID,
+    * suitable for embedding in a ChangeMessage.  This only
+    * embeds the user id since embedding additional user
+    * info could become stale if it later changes.
+    * Naturally, this id should be replaced with appropriate
+    * user data on retrieval before being dispalyed.  The
+    * formatUserRE() method can help with this task.
+    */
+  public static String formatUser(Account.Id id) {
+    return "[[user: " + id + "]]";
+  }
+}
-- 
1.7.8

